xarray/core/dataset.py

```diff
@@ -1491,7 +1491,15 @@ def __setitem__(self, key: Hashable, value) -> None:
                 "cannot yet use a dictionary as a key to set Dataset values"
             )
 
-        self.update({key: value})
+        if isinstance(key, list):
+            if len(key) != len(value):
+                raise ValueError("keys and value must have the same length.")
+            if isinstance(value, Dataset):
+                self.update(dict(zip(key, value.data_vars.values())))
```

Comment:
Is this correct? If the `key` and `data_vars` are out of order, is it going to be doing the right thing? It may be OK to exclude this option for the moment.

Reply:
For the case:
```python
ds[['foo', 'bar']] = another_ds[['bar', 'foo']]
```
it could be safe to assume the user wants to swap variable names since this is quite explicit. Technically there could be a problem with:
```python
ds[['foo', 'bar']] = another_ds 
```
where `another_ds` contains only the variables `foo` and `bar` but in the wrong order. Maybe such cases can be identified and a warning issued (which reminds the user it's better to be explicit as in the first example)

Reply:
Good point — even in the first case it's still a dataset, rather than a list of variables.

It does seem a bit dangerous but also that it's not _that_ easy to do accidentally.

Any thoughts from others?